Secp256k1.xs

```diff
@@ -626,6 +675,117 @@ _sign_schnorr(self, privkey, message)
 
 		secp256k1_perl_replace_schnorr_signature(ctx, result_signature);
 
+# Signs a recoverable digest
+void
+_sign_recoverable(self, privkey, message)
+		SV* self
+		SV* privkey
+		SV* message
+	CODE:
+		secp256k1_perl *ctx = ctx_from_sv(self);
+
+		unsigned char *message_str = size_bytestr_from_sv(message, CURVE_SIZE, "digest");
+		unsigned char *seckey_str = size_bytestr_from_sv(privkey, CURVE_SIZE, "private key");
+
+		secp256k1_ecdsa_recoverable_signature *result_signature = malloc(sizeof *result_signature);
+		int result = secp256k1_ecdsa_sign_recoverable(
+			ctx->ctx,
+			result_signature,
+			message_str,
+			seckey_str,
+			NULL,
+			NULL
+		);
+
+		if (!result) {
+			free(result_signature);
+			croak("signing failed (nonce generation problem?)");
+		}
+
+		secp256k1_perl_replace_recoverable_signature(ctx, result_signature);
+
+# Serializes recoverable signature to compact format (64 bytes + recovery_id)
+SV*
+_serialize_compact_recoverable(self, recoverable_signature)
+		SV* self
+		SV* recoverable_signature
+	CODE:
+		secp256k1_perl *ctx = ctx_from_sv(self);
+
+		if (!SvOK(recoverable_signature) || SvROK(recoverable_signature)) {
+			croak("recoverable signature must be defined and not a reference");
+		}
+
+		STRLEN sig_size;
+		unsigned char *sig_data = bytestr_from_sv(recoverable_signature, &sig_size);
+
+		if (sig_size != RECOVERABLE_SIGNATURE_SIZE) {
+			croak("invalid recoverable signature size");
+		}
+
+        secp256k1_ecdsa_recoverable_signature *signature = (secp256k1_ecdsa_recoverable_signature*) sig_data;
+
+		unsigned char signature_data[64];
+		int recovery_id;
+
+		int result = secp256k1_ecdsa_recoverable_signature_serialize_compact(
+			ctx->ctx,
+			signature_data,
+			&recovery_id,
+			signature
+		);
+
+		if (!result) {
+			croak("failed to serialize recoverable signature");
+		}
+
+		HV *return_hash = newHV();
+		hv_stores(return_hash, "signature", newSVpv((char*) signature_data, 64));
+		hv_stores(return_hash, "recovery_id", newSViv(recovery_id));
+
+		RETVAL = newRV_noinc((SV*) return_hash);
+	OUTPUT:
+		RETVAL
+
+# Recover an ECDSA public key from a recoverable signature and message hash
+void
+_recover_pubkey_recoverable(self, recoverable_signature, message_hash)
+		SV* self
+		SV* recoverable_signature
```

Comment:
This misses the point of low / high level API distinction. The signature should be passed in and out of XS through `_signature_recoverable` once you have implemented it. Having it as an argument here is a duplication in the low level API

Reply:
I've changed it to use the signature over `_signature_recoverable`, I was trying to make `recover_pubkey` to not depend on it, but it does makes sense this way